File: #0015.3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

        sort(nums.begin(),nums.end());
        vector <vector<int>> result;
        int size = nums.size();
        int low;
        int high;
        int sum;
        for (int i = 0; i < size - 2; i++) {
            if (i == 0 || (i > 0 && nums[i] != nums[i-1])) {
                low = i+1;
                high = size -1;
                sum = 0 - nums[i];
                while (low < high) {
                    if (nums[low] + nums[high] == sum) {
                        result.push_back({nums[i],nums[low],nums[high]});
                        while (low < high && nums[low] == nums[low+1]) {
                            low++;
                        }
                        while (low < high && nums[high] == nums[high-1]) {
                            high--;
                        }
                        low++;
                        high--;
                    } else if(nums[low] + nums[high] > sum) {
                        high--;
                    } else {
                        low++;
                    }
                }
            }
        }
        return result;
    }
};

File: #0015.3sum.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::iter::FromIterator;

impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {

        let mut result: HashSet<Vec<i32>> = HashSet::new();

        for (i, x) in nums.iter().enumerate() {
            let target = 0 - *x;
            let mut hm: HashMap<i32, i32> = HashMap::new();
            for y in &nums[i + 1..] {
                match hm.get(y) {
                    Some(&z) => {
                        let mut v = vec![*x, *y, z];
                        v.sort();
                        result.insert(v);
                    }
                    None => {
                        hm.insert(target - *y, *y);
                    }
                }
            }
        }
        return Vec::from_iter(result.into_iter())
    }
}
